### PR TITLE
Remove PVC's selector in backup's PVC action.

### LINKED ISSUE #6444: CSI plugin cannot work with PVC using selector to filter PV
**Describe the problem/challenge you have**

When PV and PVC are already dynamically allocated, do the following operation to make the PVC using a `selector` to filter the PV. Then create a Velero backup with the Velero CSI plugin, and restore from the backup. The restore will success, but the PVC and related Pod are always in a `Pending` state due to no available PV can be found.

* Change the Reclaim policy to Retain: `kubectl patch pv <your-pv-name> -p '{"spec":{"persistentVolumeReclaimPolicy":"Retain"}}'`
*  Tag the PV: `kubectl label pv <pv-name> <key=pair>`
* Delete the PVC and modify it to use labels to invoke the PV just labeled
* Change the PV to Available state: `kubectl patch pv <pv-name> -p '{"spec":{"claimRef": null}}'`
* Deploy the PVC modified to invoke PV using labels.

``` yaml
# k -n restore describe pvc my-pvc
Name:          my-pvc
Namespace:     restore
StorageClass:  standard-rwo
Status:        Pending
Volume:        
Labels:        app=hello-app
               backup.velero.io/must-include-additional-items=true
               velero.io/backup-name=3240504-test
               velero.io/restore-name=3240504-test-20230630061156
               velero.io/volume-snapshot-name=velero-my-pvc-6qshv
Annotations:   backup.velero.io/must-include-additional-items: true
               velero.io/backup-name: 3240504-test
               velero.io/volume-snapshot-name: velero-my-pvc-6qshv
               volume.beta.kubernetes.io/storage-provisioner: pd.csi.storage.gke.io
               volume.kubernetes.io/selected-node: <node-name>
               volume.kubernetes.io/storage-provisioner: pd.csi.storage.gke.io
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:      
Access Modes:  
VolumeMode:    Filesystem
DataSource:
  APIGroup:  snapshot.storage.k8s.io
  Kind:      VolumeSnapshot
  Name:      velero-my-pvc-6qshv
Used By:     hello-app-7579d6745b-nzjlx
Events:
  Type     Reason                Age                From                                                                                              Message
  ----     ------                ----               ----                                                                                              -------
  Normal   WaitForFirstConsumer  37s (x2 over 37s)  persistentvolume-controller                                                                       waiting for first consumer to be created before binding
  Normal   Provisioning          6s (x6 over 37s)   External provisioner is provisioning volume for claim "restore/my-pvc"
  Warning  ProvisioningFailed    6s (x6 over 37s)  failed to provision volume with StorageClass "standard-rwo": claim Selector is not supported
  Normal   ExternalProvisioning  3s (x4 over 37s)   persistentvolume-controller                                                                       waiting for a volume to be created, either by external provisioner "pd.csi.storage.gke.io" or manually created by system administrator
```

**Describe the solution you'd like**

Although this issue is triggered by a hacking way to create PV and PVC, need to consider whether there is a way to make CSI restore in the scenario.

**Anything else you would like to add:**

**Environment:**

- Velero version (use `velero version`): v1.12
- Kubernetes version (use `kubectl version`): v1.25
- Kubernetes installer & version: GKE
- Cloud provider or hardware configuration: GCP
- OS (e.g. from `/etc/os-release`): Debian

**Vote on this issue!**

This is an invitation to the Velero community to vote on issues, you can see the project's [top voted issues listed here](https://github.com/vmware-tanzu/velero/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).  
Use the "reaction smiley face" up to the right of this comment to vote.

- :+1: for "The project would be better with this feature added"
- :-1: for "This feature will not enhance the project in a meaningful way"

### PR DESCRIPTION
**Pull request by blackpiglet**
Thank you for contributing to Velero!

# Please add a summary of your change

# Does your change fix a particular issue?

Fixes #6444 

# Please indicate you've done the following:

- [x] [Accepted the DCO](https://velero.io/docs/v1.5/code-standards/#dco-sign-off). Commits without the DCO will delay acceptance.
- [x] [Created a changelog file](https://velero.io/docs/v1.5/code-standards/#adding-a-changelog) or added `/kind changelog-not-required` as a comment on this pull request.
- [ ] Updated the corresponding documentation in `site/content/docs/main`.

### CODE REVIEW

**Review by Lyndon-Li — APPROVED** [review]

**Comment by reasonerjt** [comment]
Can we make sure the removal of labelSelector only happens when the restore will be done via dynamic provision? (podvolume, CSI).

**Review by Lyndon-Li — DISMISSED** [review]

**Review by reasonerjt — DISMISSED** [review]
Approving the PR to unblock merging it, I'm fine the cosmetic code change be applied afterwards

File: pkg/backup/backup_pv_action.go [hunk 1]
```diff
@@ -72,11 +72,21 @@ func (a *PVCAction) Execute(item runtime.Unstructured, backup *v1.Backup) (runti
...
-	// remove label selectors with "velero.io/" prefixing in the key which is left by Velero restore
-	if pvc.Spec.Selector != nil && pvc.Spec.Selector.MatchLabels != nil {
-		for k := range pvc.Spec.Selector.MatchLabels {
-			if strings.HasPrefix(k, "velero.io/") {
-				delete(pvc.Spec.Selector.MatchLabels, k)
+	if pvc.Spec.Selector != nil {
+		// When StorageClassName is set to "", it means no StorageClass is specified,
+		// even the default StorageClass is not used. Only keep the Selector for this case.
+		// https://kubernetes.io/docs/concepts/storage/persistent-volumes/#reserving-a-persistentvolume
+		if pvc.Spec.StorageClassName == nil || (pvc.Spec.StorageClassName != nil && *pvc.Spec.StorageClassName != "") {
```
[the remark this answers is not in the document]
Comment (blackpiglet) [reply pkg/backup/backup_pv_action.go:79]: Thanks. Will modify.

File: pkg/backup/backup_pv_action.go [hunk 2]
```diff
@@ -72,11 +72,21 @@ func (a *PVCAction) Execute(item runtime.Unstructured, backup *v1.Backup) (runti
...
+			// Clean the selector to make the PVC to dynamically allocate PV.
+			pvc.Spec.Selector = nil
+		}
+
+		// remove label selectors with "velero.io/" prefixing in the key which is left by Velero restore
+		if pvc.Spec.Selector != nil && pvc.Spec.Selector.MatchLabels != nil {
```
[the remark this answers is not in the document]
Comment (blackpiglet) [reply pkg/backup/backup_pv_action.go:85]: The `pvc.Spec.Selector` can be set to nil in the preceding logic, so it needs to be checked again.

``` go
	if pvc.Spec.Selector != nil {
		// When StorageClassName is set to "", it means no StorageClass is specified,
		// even the default StorageClass is not used. Only keep the Selector for this case.
		// https://kubernetes.io/docs/concepts/storage/persistent-volumes/#reserving-a-persistentvolume
		if pvc.Spec.StorageClassName == nil || *pvc.Spec.StorageClassName != "" {
			// Clean the selector to make the PVC to dynamically allocate PV.
			pvc.Spec.Selector = nil
		}
    ......
```

**Review by Lyndon-Li — DISMISSED** [review]

File: pkg/backup/backup_pv_action.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (blackpiglet) [reply pkg/backup/backup_pv_action.go]: Modified.

**Review by reasonerjt — APPROVED** [review]

**Review by Lyndon-Li — APPROVED** [review]

**Pull request merged**